From 0572ad5f568b4739dc6436469458ba018621a576 Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Fri, 28 Mar 2025 09:54:27 +0545
Subject: [PATCH 14/15] Introduce PropImitationHooks

Squashed commit of the following:

commit 0f22c4b8e13eb5b52f97f57092eba7efc7f3dd43
Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Sat Dec 2 22:58:20 2023 +0530

    PIHooks: Add a system prop to enable debug logs

    Change-Id: I949b21a5a10c575eede42e61f5f75bdbefd9f161

commit 65cfc82416cd4bd6f576bbf381f5816b14b7c3b6
Author: Jyotiraditya Panda <jyotiraditya@aospa.co>
Date:   Wed Aug 7 21:35:48 2024 +0530

    PIHooks: Allow user to disable prop imitation

     - Prevents spoofing props, bypassing task perms and blocking key attestation for GMS.

    [adithya2306: add a prop for disabling only key attestation too]

    Change-Id: I529e91e8d11a27d064c3ef85a8cf5e73e1857e3b
    Signed-off-by: Jyotiraditya Panda <jyotiraditya@aospa.co>
    Co-authored-by: Adithya R <gh0strider.2k18.reborn@gmail.com>

commit 4e39f4db7396ead1d2703eb276f2e090d09a6f36
Author: bheatleyyy <sugakesshaa@pixelos.net>
Date:   Sun Dec 29 20:49:52 2024 +0100

    PIHooks: Simplify tablet spoofing

    Change-Id: Iec15229e2a481f6c0d9bfb560780ca95caf961db

commit 06c17aa8579224be1bf511267f92b90c82f268a8
Author: basamaryan <basam.aryan@gmail.com>
Date:   Tue Jul 2 13:14:34 2024 +0200

    PIHooks: Refactor

    - Rearrange apps and processes to be spoofed alphabetically
    - Drop Play store and Turbo spoofing
    - Spoof the Pixel Launcher to husky
    - Spoof a few GMS processes to barbet
    - Switch from sailfish to marlin for Google Photos spoofing
    - Add a constant set for Nexus and Tensor features
    - Drop config_spoofGoogleApps, as we would always want this to be true
    - Move to using switch cases for spoofing
    - Switch to caiman, and update fingerprints.

    This more closely aligns with https://github.com/TheParasiteProject/vendor_parasite_common/blob/bf6cfe2320472c15a70c84df42d55ec366ffe8d6/libs/com.android.internal.util.custom/src

    Co-authored-by: Hridaya Prajapati <info.hridayaprajapati@gmail.com>
    Change-Id: I61662a1d89ccea4776d78e2c9446b7c1e75314ba

commit 628f7d99e151cd0c71a7d341e722496976078fbb
Author: someone5678 <nemui3353@gmail.com>
Date:   Sun Aug 6 19:29:27 2023 +0900

    PIHooks: Set HARDWARE, ID values

    * Google apps also check these values

    Change-Id: I846986121faec39d3ece044794fd0f459e8d435d

commit 59f7f69419ff9e0f855d27721d7510b93f9a3f64
Author: basamaryan <basam.aryan@gmail.com>
Date:   Tue Jun 4 20:46:55 2024 +0200

    PIHooks: Drop GBoard spoofing

    Change-Id: Ibc567d79f4887d7db471cfca496731fa726be543

commit 03ccebe68f3d1af7b94ac7354fba08246009a645
Author: saikiran2001 <bjsaikiran@gmail.com>
Date:   Fri Dec 29 07:52:06 2023 +0000

    PIHooks: Spoof as Pixel Tab for tablets

    Change-Id: I2d5c60dc144d56ed1957c5e2750fc60831bf2334

commit c3821a210d7a684cd392c2306bbbfe4d490edf4b
Author: PainKiller3 <ninadpatil100@gmail.com>
Date:   Sat Dec 9 18:20:08 2023 +0530

    PIHooks: Spoof more Google apps

    Change-Id: I9faf3b3ffeba74130d9af77f191eae46b44f4b40

commit 34d7ce8582723a736a1eb7c188cec2e485c9c2ba
Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Mon Sep 19 01:07:09 2022 +0530

    PIHooks: Spoof Pixel 8 Pro to some Google apps

    This unlocks some non-TPU reliant, useful features across all devices.

    Co-authored-by: Sourajit Karmakar <sourajit.karmakar@gmail.com>
    Change-Id: Ieb09625d5d0d2d5f7036db4ae5a0036b9e4a48fa
    Signed-off-by: ralph950412 <ralph950412@gmail.com>
    Signed-off-by: PainKiller3 <ninadpatil100@gmail.com>

commit 8d8f53d15eddba3193ee4c1cac92985e470cef9b
Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Thu Feb 9 10:51:56 2023 +0530

    PIHooks: Spoof more gapps and google photos

    Change-Id: I2e66cba7c9098dd0063123db103407b2955d1c80

commit e3dee306b9bbe39db8a92637c5630ba93292594e
Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Sun Sep 1 09:46:29 2024 +0530

    PIHooks: Make it a hidden API

    Change-Id: I271bc644444581565847d56766ca85bc7d272436

commit 12d6e8ed3ec94b2b4df8c69680830b41c28b9c52
Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Fri Dec 8 05:59:39 2023 +0530

    PIHooks: Set system props for native hook

    Relay the spoofed (or not) first API and security patch level to
    native PIHooks.

    Change-Id: I11a9e9c13a33c1f9962844abba7c83bd040d7a62

commit 47c2ba7d229c925b88bd23ddc6bb7922b246eaa7
Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Sat Dec 2 23:43:18 2023 +0530

    PIHooks: Get resources from context

    Allows updating overlays on the fly.

    Change-Id: Idb85918af06400559e3b17317b3ab88a25828fae

commit f384b2d2a4a6664352188fe4b2a541a831c5c4c1
Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Sat Nov 18 20:08:08 2023 +0530

    PIHooks: Allow setting any field in certified build props

    Brand and manufacturer may also be used for attestation checks, as seen
    by _FOR_ATTESTATION fields added in Android 14 [1]. Refactor certified
    properties to allow setting any Build or Build.VERSION string or integer
    field, to make it more scalable for future.

    [1]: https://cs.android.com/android/_/android/platform/frameworks/base/+/fb32aac1ea60bf267ff013f6e1023641736c6de5

    Change-Id: I16dc8e18a67ffd8fdba8eb9c6928447f82b7d810

commit e7e8637f2c634cdda9fa0d74c17a03a3bd9ef765
Author: Omkar Chandorkar <gotenksIN@aospa.co>
Date:   Thu Sep 1 17:18:58 2022 +0530

    Introduce PropImitationHooks

    This is a squash of the following commits.

      Author: Omkar Chandorkar <gotenksIN@aospa.co>
      Date:   Thu Sep 1 17:18:58 2022 +0530

          Introduce PropImitationHooks

          - A minimal implementation to pass CTS and Play Integrity while
            loading correct ARCore profiles if existing

          - Loosely inspired by PixelPropUtils from
            https://github.com/PixelExperience/frameworks_base/blob/twelve/core/java/com/android/internal/util/custom/PixelPropsUtils.java

          - Thanks Danny for original idea at
            https://github.com/ProtonAOSP/android_frameworks_base/blob/sc-qpr3/core/java/com/android/internal/gmscompat/AttestationHooks.java

          - Thanks Dyneteve for fixing Play Integrity API at
            https://github.com/hentaiOS/platform_frameworks_base/commit/c9a086f9b635bf4e3e61fbd501ade464f8f5b713

          Change-Id: I6937258c8297aa909fcdb3b40195cbd6dc0f869b
          Co-authored-by: Chris Crawford <chriscrawford893@gmail.com>
          Co-authored-by: jhenrique09 <jhenrique09.mcz@hotmail.com>
          Co-authored-by: Danny Lin <danny@kdrag0n.dev>
          Co-authored-by: Adithya R <gh0strider.2k18.reborn@gmail.com>
          Signed-off-by: Omkar Chandorkar <gotenksIN@aospa.co>

      Author: Danny Lin <danny@kdrag0n.dev>
      Date:   Fri Oct 14 09:53:12 2022 +0200

          PIHooks: Alter model name to avoid SafetyNet HW attestation

          As of September 2, Google is enforcing SafetyNet's previously
          opportunistic hardware-backed attestation based on device information.
          Append a space to the device model name in order to avoid such
          enforcement.

          Change-Id: Ib7779e0aae40cab3730a56785e9231896917ab0a
          [ghostrider-reborn: adapt to PIHooks, use invisible space]
          Signed-off-by: Adithya R <gh0strider.2k18.reborn@gmail.com>

      Author: jhenrique09 <jhsv09@gmail.com>
      Date:   Fri Nov 4 15:40:04 2022 -0300

          PIHooks: Get package name from context instead of app

          11-04 08:48:39.039 11637 11637 E AndroidRuntime: FATAL EXCEPTION: main
          11-04 08:48:39.039 11637 11637 E AndroidRuntime: Process: com.NextFloor.DestinyChild, PID: 11637
          11-04 08:48:39.039 11637 11637 E AndroidRuntime: java.lang.RuntimeException: Unable to get provider androidx.startup.InitializationProvider: androidx.startup.StartupException: android.content.pm.PackageManager$NameNotFoundException: ComponentInfo{/androidx.startup.InitializationProvider}
          11-04 08:48:39.039 11637 11637 E AndroidRuntime:   at android.app.ActivityThread.installProvider(ActivityThread.java:7488)

          Change-Id: Icb12f938fe0fca710f8f9d29182d0134ba3c63eb
          [ghostrider-reborn: adapt to PIHooks]
          Signed-off-by: Adithya R <gh0strider.2k18.reborn@gmail.com>

      Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
      Date:   Thu Feb 9 11:41:44 2023 +0530

          PIHooks: Make CTS/Play Integrity pass again

          The logic behind SafetyNet CTS and Play Integrity has been updated today
          it now checks the product and model names against the fingerprint and if
          they do not match the CTS profile will fail.

          Based on https://github.com/hentaiOS/platform_frameworks_base/commit/32cd021753626ae2dc0f72d222ac4fd734f3a8aa

          Co-authored-by: Dyneteve <dyneteve@hentaios.com>
          Change-Id: If3a865849c7c99a9c7080114ba94a8f8878bb4f2

      Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
      Date:   Thu Jun 8 16:55:44 2023 +0530

          PIHooks: Un-spoof GMS while adding google account

          GMS uses its DroidGuard service (the same that's used for safetynet)
          to get the device name from the build fingerprint, which from our
          current certified props, is Pixel XL. The device name after Google
          login shows up in the login alert accordingly as Pixel XL, which is
          inconvenient especially when you have multiple devices on the same ROM.

          By restarting the DroidGuard service process (gms.unstable) as soon as
          the user starts adding Google account (i.e when the GMS add account
          activity is started and moved to the top of the task stack), we can
          skip spoofing to certified props in this instance. When the user
          stops/finishes adding account (GMS add account activity is no longer
          at the top of the task stack), restart the process again, this time
          spoofing to certified props in order to be ready for safetynet.

          This ensures while adding a Google account, GMS reads the original
          properties of the device and labels the device with its original name,
          while in any other situation GMS reads certified properties required
          to pass SafetyNet/Play Integrity.

          Test:
          - Add google account from setup wizard, settings, any other app
          - Observe correct device name in gmail login alert
          - SafetyNet (basic, CTS), Play Integrity (basic, device) passes.

          Change-Id: Ia5463f220a8262f7963fc5b2640a7a5311a73210

      Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
      Date:   Mon Sep 19 17:33:34 2022 +0530

          PIHooks: Allow spoofing device model for Netflix

          It can be used to unlock HDR for devices that don't
          have it by default.

          Set config_netflixSpoofModel to the desired model in
          device overlay.

          Change-Id: I5d520de143c16e9c6fd10978244f598bd64b4d0a

    Change-Id: I7844dc9ac8e2f927e6bb565ee218210b194e6ef0

Change-Id: Ib0be154cb27e5c67233564278727a0b093ab1cac
---
 .../app/ApplicationPackageManager.java        |   4 +-
 core/java/android/app/Instrumentation.java    |   3 +
 .../internal/util/PropImitationHooks.java     | 408 ++++++++++++++++++
 core/res/res/values/custom_config.xml         |  22 +
 core/res/res/values/custom_symbols.xml        |   7 +
 .../keystore2/AndroidKeyStoreSpi.java         |   3 +
 .../server/wm/ActivityTaskManagerService.java |   9 +-
 7 files changed, 453 insertions(+), 3 deletions(-)
 create mode 100644 core/java/com/android/internal/util/PropImitationHooks.java
 create mode 100644 core/res/res/values/custom_config.xml
 create mode 100644 core/res/res/values/custom_symbols.xml

diff --git a/core/java/android/app/ApplicationPackageManager.java b/core/java/android/app/ApplicationPackageManager.java
index 8da6f8605ca4..79ae3cd5ba89 100644
--- a/core/java/android/app/ApplicationPackageManager.java
+++ b/core/java/android/app/ApplicationPackageManager.java
@@ -133,6 +133,7 @@ import com.android.internal.annotations.GuardedBy;
 import com.android.internal.annotations.VisibleForTesting;
 import com.android.internal.os.SomeArgs;
 import com.android.internal.pm.RoSystemFeatures;
+import com.android.internal.util.PropImitationHooks;
 import com.android.internal.util.UserIcons;
 
 import com.nvidia.NvAppProfileService;
@@ -855,7 +856,8 @@ public class ApplicationPackageManager extends PackageManager {
                 return maybeHasSystemFeature;
             }
         }
-        return mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version));
+        return PropImitationHooks.hasSystemFeature(name,
+        mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version)));
     }
 
     /** @hide */
diff --git a/core/java/android/app/Instrumentation.java b/core/java/android/app/Instrumentation.java
index 19fecb9bf7c2..f1b20dd1f586 100644
--- a/core/java/android/app/Instrumentation.java
+++ b/core/java/android/app/Instrumentation.java
@@ -65,6 +65,7 @@ import android.view.Window;
 import android.view.WindowManagerGlobal;
 
 import com.android.internal.content.ReferrerIntent;
+import com.android.internal.util.PropImitationHooks;
 
 import java.io.File;
 import java.lang.annotation.Retention;
@@ -1346,6 +1347,7 @@ public class Instrumentation {
         Application app = getFactory(context.getPackageName())
                 .instantiateApplication(cl, className);
         app.attach(context);
+        PropImitationHooks.setProps(context);
         return app;
     }
     
@@ -1363,6 +1365,7 @@ public class Instrumentation {
             ClassNotFoundException {
         Application app = (Application)clazz.newInstance();
         app.attach(context);
+        PropImitationHooks.setProps(context);
         return app;
     }
 
diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
new file mode 100644
index 000000000000..b95ba943137f
--- /dev/null
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -0,0 +1,408 @@
+/*
+ * Copyright (C) 2022 Paranoid Android
+ *           (C) 2023 ArrowOS
+ *           (C) 2023 The LibreMobileOS Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.internal.util;
+
+import android.app.ActivityTaskManager;
+import android.app.Application;
+import android.app.TaskStackListener;
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.res.Resources;
+import android.os.Build;
+import android.os.Binder;
+import android.os.Process;
+import android.os.SystemProperties;
+import android.text.TextUtils;
+import android.util.Log;
+
+import com.android.internal.R;
+
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @hide
+ */
+public class PropImitationHooks {
+
+    private static final String TAG = "PropImitationHooks";
+    private static final boolean DEBUG = SystemProperties.getBoolean("debug.pihooks.log", false);
+
+    private static final String PACKAGE_AIWALLPAPERS = "com.google.android.apps.aiwallpapers";
+    private static final String PACKAGE_ARCORE = "com.google.ar.core";
+    private static final String PACKAGE_ASI = "com.google.android.as";
+    private static final String PACKAGE_ASSISTANT = "com.google.android.apps.googleassistant";
+    private static final String PACKAGE_EMOJIWALLPAPER = "com.google.android.apps.emojiwallpaper";
+
+    private static final String PACKAGE_FINSKY = "com.android.vending";
+    private static final String PACKAGE_GMS = "com.google.android.gms";
+    private static final String PACKAGE_GPHOTOS = "com.google.android.apps.photos";
+    private static final String PACKAGE_NETFLIX = "com.netflix.mediaclient";
+
+    private static final String PACKAGE_NEXUSLAUNCHER = "com.google.android.apps.nexuslauncher";
+    private static final String PACKAGE_PIXELTHEMES = "com.google.android.apps.customization.pixel";
+    private static final String PACKAGE_PIXELWALLPAPER = "com.google.android.apps.wallpaper.pixel";
+    private static final String PACKAGE_LIVEWALLPAPER = "com.google.pixel.livewallpaper";
+    private static final String PACKAGE_SUBSCRIPTION_RED = "com.google.android.apps.subscriptions.red";
+    private static final String PACKAGE_VELVET = "com.google.android.googlequicksearchbox";
+    private static final String PACKAGE_WALLPAPER = "com.google.android.apps.wallpaper";
+    private static final String PACKAGE_WALLPAPEREFFECTS = "com.google.android.wallpaper.effects";
+
+    private static final String PROCESS_GMS_GAPPS = PACKAGE_GMS + ".gapps";
+    private static final String PROCESS_GMS_GSERVICE = PACKAGE_GMS + ".gservice";
+    private static final String PROCESS_GMS_LEARNING = PACKAGE_GMS + ".learning";
+    private static final String PROCESS_GMS_PERSISTENT = PACKAGE_GMS + ".persistent";
+    private static final String PROCESS_GMS_SEARCH = PACKAGE_GMS + ".search";
+    private static final String PROCESS_GMS_UNSTABLE = PACKAGE_GMS + ".unstable";
+    private static final String PROCESS_GMS_UPDATE = PACKAGE_GMS + ".update";
+
+    private static final String PROP_SECURITY_PATCH = "persist.sys.pihooks.security_patch";
+    private static final String PROP_FIRST_API_LEVEL = "persist.sys.pihooks.first_api_level";
+
+    private static final ComponentName GMS_ADD_ACCOUNT_ACTIVITY = ComponentName.unflattenFromString(
+            "com.google.android.gms/.auth.uiflows.minutemaid.MinuteMaidActivity");
+
+    private static final Boolean sDisableGmsProps = SystemProperties.getBoolean(
+            "persist.sys.pihooks.disable.gms_props", false);
+
+    private static final Boolean sDisableKeyAttestationBlock = SystemProperties.getBoolean(
+        "persist.sys.pihooks.disable.gms_key_attestation_block", false);
+
+    private static final Map<String, String> sPixelNineProps = Map.of(
+            "PRODUCT", "caiman",
+            "DEVICE", "caiman",
+            "HARDWARE", "caiman",
+            "MANUFACTURER", "Google",
+            "BRAND", "google",
+            "MODEL", "Pixel 9 Pro",
+            "ID", "AD1A.240905.004",
+            "FINGERPRINT", "google/caiman/caiman:14/AD1A.240905.004/12196292:user/release-keys"
+    );
+
+    private static final Map<String, String> sPixelFiveProps = Map.of(
+            "PRODUCT", "barbet",
+            "DEVICE", "barbet",
+            "HARDWARE", "barbet",
+            "MANUFACTURER", "Google",
+            "BRAND", "google",
+            "MODEL", "Pixel 5a",
+            "ID", "AP2A.240805.004",
+            "FINGERPRINT", "google/barbet/barbet:14/AP2A.240805.005/12025142:user/release-keys"
+    );
+
+    private static final Map<String, String> sPixelTabletProps = Map.of(
+            "PRODUCT", "tangorpro",
+            "DEVICE", "tangorpro",
+            "HARDWARE", "tangorpro",
+            "MANUFACTURER", "Google",
+            "BRAND", "google",
+            "MODEL", "Pixel Tablet",
+            "ID", "AP2A.240905.003",
+            "FINGERPRINT", "google/tangorpro/tangorpro:14/AP2A.240905.003/12231197:user/release-keys"
+    );
+
+    private static final Map<String, String> sPixelXLProps = Map.of(
+            "PRODUCT", "marlin",
+            "DEVICE", "marlin",
+            "HARDWARE", "marlin",
+            "MANUFACTURER", "Google",
+            "BRAND", "google",
+            "MODEL", "Pixel XL",
+            "ID", "QP1A.191005.007.A3",
+            "FINGERPRINT", "google/marlin/marlin:10/QP1A.191005.007.A3/5972272:user/release-keys"
+    );
+
+    private static final Set<String> sNexusFeatures = Set.of(
+            "NEXUS_PRELOAD",
+            "nexus_preload",
+            "GOOGLE_BUILD",
+            "GOOGLE_EXPERIENCE",
+            "PIXEL_EXPERIENCE"
+    );
+
+    private static final Set<String> sPixelFeatures = Set.of(
+            "PIXEL_2017_EXPERIENCE",
+            "PIXEL_2017_PRELOAD",
+            "PIXEL_2018_EXPERIENCE",
+            "PIXEL_2018_PRELOAD",
+            "PIXEL_2019_EXPERIENCE",
+            "PIXEL_2019_MIDYEAR_EXPERIENCE",
+            "PIXEL_2019_MIDYEAR_PRELOAD",
+            "PIXEL_2019_PRELOAD",
+            "PIXEL_2020_EXPERIENCE",
+            "PIXEL_2020_MIDYEAR_EXPERIENCE",
+            "PIXEL_2021_MIDYEAR_EXPERIENCE"
+    );
+
+    private static final Set<String> sTensorFeatures = Set.of(
+            "PIXEL_2021_EXPERIENCE",
+            "PIXEL_2022_EXPERIENCE",
+            "PIXEL_2022_MIDYEAR_EXPERIENCE",
+            "PIXEL_2023_EXPERIENCE",
+            "PIXEL_2023_MIDYEAR_EXPERIENCE",
+            "PIXEL_2024_EXPERIENCE",
+            "PIXEL_2024_MIDYEAR_EXPERIENCE"
+    );
+
+    private static volatile String[] sCertifiedProps;
+    private static volatile String sStockFp, sNetflixModel;
+
+    private static volatile String sProcessName;
+    private static volatile boolean sIsGms, sIsFinsky, sIsPhotos;
+
+    public static void setProps(Context context) {
+        final String packageName = context.getPackageName();
+        final String processName = Application.getProcessName();
+
+        if (TextUtils.isEmpty(packageName) || TextUtils.isEmpty(processName)) {
+            Log.e(TAG, "Null package or process name");
+            return;
+        }
+
+        final Resources res = context.getResources();
+        if (res == null) {
+            Log.e(TAG, "Null resources");
+            return;
+        }
+
+        sCertifiedProps = res.getStringArray(R.array.config_certifiedBuildProperties);
+        sStockFp = res.getString(R.string.config_stockFingerprint);
+        sNetflixModel = res.getString(R.string.config_netflixSpoofModel);
+
+        sProcessName = processName;
+        sIsGms = packageName.equals(PACKAGE_GMS) && processName.equals(PROCESS_GMS_UNSTABLE);
+        sIsFinsky = packageName.equals(PACKAGE_FINSKY);
+        sIsPhotos = packageName.equals(PACKAGE_GPHOTOS);
+
+        /* Set certified properties for GMSCore
+         * Set stock fingerprint for ARCore
+         * Set Pixel 8 Pro for Google, ASI and GMS device configurator
+         * Set Pixel XL for Google Photos
+         * Set custom model for Netflix
+         */
+
+        switch (processName) {
+            case PROCESS_GMS_UNSTABLE:
+                dlog("Setting certified props for: " + packageName + " process: " + processName);
+                setCertifiedPropsForGms();
+                return;
+            case PROCESS_GMS_PERSISTENT:
+            case PROCESS_GMS_GAPPS:
+            case PROCESS_GMS_GSERVICE:
+            case PROCESS_GMS_LEARNING:
+            case PROCESS_GMS_SEARCH:
+            case PROCESS_GMS_UPDATE:
+                dlog("Spoofing Pixel 5a for: " + packageName + " process: " + processName);
+                setProps(sPixelFiveProps);
+                return;
+        }
+
+        if (!sStockFp.isEmpty() && packageName.equals(PACKAGE_ARCORE)) {
+            dlog("Setting stock fingerprint for: " + packageName);
+            setPropValue("FINGERPRINT", sStockFp);
+            return;
+        }
+
+        switch (packageName) {
+            case PACKAGE_AIWALLPAPERS:
+            case PACKAGE_ASSISTANT:
+            case PACKAGE_ASI:
+            case PACKAGE_EMOJIWALLPAPER:
+            case PACKAGE_GMS:
+            case PACKAGE_LIVEWALLPAPER:
+            case PACKAGE_NEXUSLAUNCHER:
+            case PACKAGE_PIXELTHEMES:
+            case PACKAGE_PIXELWALLPAPER:
+            case PACKAGE_SUBSCRIPTION_RED:
+            case PACKAGE_VELVET:
+            case PACKAGE_WALLPAPER:
+            case PACKAGE_WALLPAPEREFFECTS:
+                if (SystemProperties.get("ro.build.characteristics").equals("tablet")) {
+                    dlog("Spoofing Pixel Tablet for: " + packageName + " process: " + processName);
+                    setProps(sPixelTabletProps);
+                } else {
+                    dlog("Spoofing Pixel 9 Pro for: " + packageName + " process: " + processName);
+                    setProps(sPixelNineProps);
+                }
+                return;
+            case PACKAGE_GPHOTOS:
+                dlog("Spoofing Pixel XL for Google Photos");
+                setProps(sPixelXLProps);
+                return;
+            case PACKAGE_NETFLIX:
+                if (!sNetflixModel.isEmpty()) {
+                    dlog("Setting model to " + sNetflixModel + " for Netflix");
+                    setPropValue("MODEL", sNetflixModel);
+                }
+                return;
+        }
+    }
+
+    private static void setProps(Map<String, String> props) {
+        props.forEach(PropImitationHooks::setPropValue);
+    }
+
+    private static void setPropValue(String key, String value) {
+        try {
+            dlog("Setting prop " + key + " to " + value.toString());
+            Class clazz = Build.class;
+            if (key.startsWith("VERSION.")) {
+                clazz = Build.VERSION.class;
+                key = key.substring(8);
+            }
+            Field field = clazz.getDeclaredField(key);
+            field.setAccessible(true);
+            // Cast the value to int if it's an integer field, otherwise string.
+            field.set(null, field.getType().equals(Integer.TYPE) ? Integer.parseInt(value) : value);
+            field.setAccessible(false);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to set prop " + key, e);
+        }
+    }
+
+    private static void setCertifiedPropsForGms() {
+        if (sDisableGmsProps) {
+            dlog("GMS prop imitation is disabled by user");
+            setSystemProperty(PROP_SECURITY_PATCH, Build.VERSION.SECURITY_PATCH);
+            setSystemProperty(PROP_FIRST_API_LEVEL,
+                    Integer.toString(Build.VERSION.DEVICE_INITIAL_SDK_INT));
+            return;
+        }
+
+        if (sCertifiedProps.length == 0) {
+            dlog("Certified props are not set");
+            return;
+        }
+        final boolean was = isGmsAddAccountActivityOnTop();
+        final TaskStackListener taskStackListener = new TaskStackListener() {
+            @Override
+            public void onTaskStackChanged() {
+                final boolean is = isGmsAddAccountActivityOnTop();
+                if (is ^ was) {
+                    dlog("GmsAddAccountActivityOnTop is:" + is + " was:" + was +
+                            ", killing myself!"); // process will restart automatically later
+                    Process.killProcess(Process.myPid());
+                }
+            }
+        };
+        if (!was) {
+            dlog("Spoofing build for GMS");
+            setCertifiedProps();
+        } else {
+            dlog("Skip spoofing build for GMS, because GmsAddAccountActivityOnTop");
+        }
+        try {
+            ActivityTaskManager.getService().registerTaskStackListener(taskStackListener);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to register task stack listener!", e);
+        }
+    }
+
+    private static void setCertifiedProps() {
+        for (String entry : sCertifiedProps) {
+            // Each entry must be of the format FIELD:value
+            final String[] fieldAndProp = entry.split(":", 2);
+            if (fieldAndProp.length != 2) {
+                Log.e(TAG, "Invalid entry in certified props: " + entry);
+                continue;
+            }
+            setPropValue(fieldAndProp[0], fieldAndProp[1]);
+        }
+        setSystemProperty(PROP_SECURITY_PATCH, Build.VERSION.SECURITY_PATCH);
+        setSystemProperty(PROP_FIRST_API_LEVEL,
+                Integer.toString(Build.VERSION.DEVICE_INITIAL_SDK_INT));
+    }
+
+    private static void setSystemProperty(String name, String value) {
+        try {
+            SystemProperties.set(name, value);
+            dlog("Set system prop " + name + "=" + value);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to set system prop " + name + "=" + value, e);
+        }
+    }
+
+    private static boolean isGmsAddAccountActivityOnTop() {
+        try {
+            final ActivityTaskManager.RootTaskInfo focusedTask =
+                    ActivityTaskManager.getService().getFocusedRootTaskInfo();
+            return focusedTask != null && focusedTask.topActivity != null
+                    && focusedTask.topActivity.equals(GMS_ADD_ACCOUNT_ACTIVITY);
+        } catch (Exception e) {
+            Log.e(TAG, "Unable to get top activity!", e);
+        }
+        return false;
+    }
+
+    public static boolean shouldBypassTaskPermission(Context context) {
+        if (sDisableGmsProps) {
+            return false;
+        }
+
+        // GMS doesn't have MANAGE_ACTIVITY_TASKS permission
+        final int callingUid = Binder.getCallingUid();
+        final int gmsUid;
+        try {
+            gmsUid = context.getPackageManager().getApplicationInfo(PACKAGE_GMS, 0).uid;
+            dlog("shouldBypassTaskPermission: gmsUid:" + gmsUid + " callingUid:" + callingUid);
+        } catch (Exception e) {
+            Log.e(TAG, "shouldBypassTaskPermission: unable to get gms uid", e);
+            return false;
+        }
+        return gmsUid == callingUid;
+    }
+
+    private static boolean isCallerSafetyNet() {
+        return sIsGms && Arrays.stream(Thread.currentThread().getStackTrace())
+                .anyMatch(elem -> elem.getClassName().contains("DroidGuard"));
+    }
+
+    public static void onEngineGetCertificateChain() {
+        if (sDisableKeyAttestationBlock) {
+            dlog("Key attestation blocking is disabled by user");
+            return;
+        }
+
+        // Check stack for SafetyNet or Play Integrity
+        if (isCallerSafetyNet() || sIsFinsky) {
+            dlog("Blocked key attestation sIsGms=" + sIsGms + " sIsFinsky=" + sIsFinsky);
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    public static boolean hasSystemFeature(String name, boolean has) {
+        if (sIsPhotos) {
+            if (has && (sPixelFeatures.stream().anyMatch(name::contains)
+                    || sTensorFeatures.stream().anyMatch(name::contains))) {
+                dlog("Blocked system feature " + name + " for Google Photos");
+                has = false;
+            } else if (!has && sNexusFeatures.stream().anyMatch(name::contains)) {
+                dlog("Enabled system feature " + name + " for Google Photos");
+                has = true;
+            }
+        }
+        return has;
+    }
+
+    public static void dlog(String msg) {
+        if (DEBUG) Log.d(TAG, "[" + sProcessName + "] " + msg);
+    }
+}
diff --git a/core/res/res/values/custom_config.xml b/core/res/res/values/custom_config.xml
new file mode 100644
index 000000000000..a7038f072ff2
--- /dev/null
+++ b/core/res/res/values/custom_config.xml
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <!-- Build properties from a GMS certified device against current platform SPL level -->
+    <string-array name="config_certifiedBuildProperties" translatable="false">
+      <!--
+        Each entry must be of the format
+            FIELD:value
+        with the field belonging to android.os.Build or android.os.Build.VERSION class.
+        Example:
+        <item>BRAND:foo</item>
+        <item>DEVICE:bar</item>
+        <item>FINGERPRINT:foo/bar/bar:1.0/lorem/ipsum:dolor/sit-amet</item>
+        <item>VERSION.RELEASE:1.0</item>
+      -->
+    </string-array>
+
+    <!-- Build fingerprint from stock ROM -->
+    <string name="config_stockFingerprint" translatable="false"></string>
+
+    <!-- Device model (Build.MODEL) to spoof for netflix, if not empty -->
+    <string name="config_netflixSpoofModel" translatable="false"></string>
+</resources>
diff --git a/core/res/res/values/custom_symbols.xml b/core/res/res/values/custom_symbols.xml
new file mode 100644
index 000000000000..eb02ba9f460f
--- /dev/null
+++ b/core/res/res/values/custom_symbols.xml
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <!-- PropImitationHooks -->
+    <java-symbol type="array" name="config_certifiedBuildProperties" />
+    <java-symbol type="string" name="config_stockFingerprint" />
+    <java-symbol type="string" name="config_netflixSpoofModel" />
+</resources>
diff --git a/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java b/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
index e6a63b9c4c17..81ca30ea8676 100644
--- a/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
+++ b/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
@@ -48,6 +48,7 @@ import android.system.keystore2.ResponseCode;
 import android.util.Log;
 
 import com.android.internal.annotations.VisibleForTesting;
+import com.android.internal.util.PropImitationHooks;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
@@ -178,6 +179,8 @@ public class AndroidKeyStoreSpi extends KeyStoreSpi {
 
     @Override
     public Certificate[] engineGetCertificateChain(String alias) {
+        PropImitationHooks.onEngineGetCertificateChain();
+
         KeyEntryResponse response = getKeyMetadata(alias);
 
         if (response == null || response.metadata.certificate == null) {
diff --git a/services/core/java/com/android/server/wm/ActivityTaskManagerService.java b/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
index fac92ee319f2..1a499e15cb4a 100644
--- a/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
+++ b/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
@@ -256,6 +256,7 @@ import com.android.internal.protolog.ProtoLog;
 import com.android.internal.util.ArrayUtils;
 import com.android.internal.util.FastPrintWriter;
 import com.android.internal.util.FrameworkStatsLog;
+import com.android.internal.util.PropImitationHooks;
 import com.android.internal.util.function.pooled.PooledLambda;
 import com.android.server.LocalManagerRegistry;
 import com.android.server.LocalServices;
@@ -2026,7 +2027,9 @@ public class ActivityTaskManagerService extends IActivityTaskManager.Stub {
 
     @Override
     public RootTaskInfo getFocusedRootTaskInfo() throws RemoteException {
-        enforceTaskPermission("getFocusedRootTaskInfo()");
+        if (!PropImitationHooks.shouldBypassTaskPermission(mContext)) {
+            enforceTaskPermission("getFocusedRootTaskInfo()");
+        }
         final long ident = Binder.clearCallingIdentity();
         try {
             synchronized (mGlobalLock) {
@@ -3187,7 +3190,9 @@ public class ActivityTaskManagerService extends IActivityTaskManager.Stub {
     /** Sets the task stack listener that gets callbacks when a task stack changes. */
     @Override
     public void registerTaskStackListener(ITaskStackListener listener) {
-        enforceTaskPermission("registerTaskStackListener()");
+        if (!PropImitationHooks.shouldBypassTaskPermission(mContext)) {
+            enforceTaskPermission("registerTaskStackListener()");
+        }
         mTaskChangeNotificationController.registerTaskStackListener(listener);
     }
 
-- 
2.50.0

